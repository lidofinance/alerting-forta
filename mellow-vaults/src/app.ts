import { getEthersProvider, fetchJwt, Finding, verifyJwt } from 'forta-agent'
import { LIDO_STETH_ADDRESS } from 'constants/common'
import { ETHProvider } from './clients/eth_provider'
import { FormatterWithEIP1898 } from './clients/eth_formatter'
import { DataRW } from './shared/mutex'
import * as Winston from 'winston'
import * as E from 'fp-ts/Either'
import { BorderTime, HealthChecker, MaxNumberErrorsPerBorderTime } from './services/health-checker/HealthChecker.srv'
import { LidoDAO__factory } from './generated'
import { VaultWatcherSrv } from './services/vault-watcher/VaultWatcher.srv'
import { MultisigWatcherSrv } from './services/multisig-watcher/MultisigWatcher.srv'

export type Container = {
  ethClient: ETHProvider
  VaultWatcherSrv: VaultWatcherSrv
  MultisigWatcherSrv: MultisigWatcherSrv
  findingsRW: DataRW<Finding>
  healthChecker: HealthChecker
}

export class App {
  private static instance: Container

  public static async getJwt(): Promise<E.Either<Error, string>> {
    let token: string
    try {
      token = await fetchJwt({})
    } catch (e) {
      return E.left(new Error(`Could not fetch jwt. Cause ${e}`))
    }

    if (process.env.NODE_ENV === 'production') {
      try {
        const isTokenOk = await verifyJwt(token)
        if (!isTokenOk) {
          return E.left(new Error(`Token verification failed`))
        }
      } catch (e) {
        return E.left(new Error(`Token verification failed`))
      }
    }

    return E.right(token)
  }

  public static async getInstance(): Promise<Container> {
    if (!App.instance) {
      const ethersProvider = getEthersProvider()
      ethersProvider.formatter = new FormatterWithEIP1898()

      const stethContract = LidoDAO__factory.connect(LIDO_STETH_ADDRESS, ethersProvider)

      const ethClient = new ETHProvider(ethersProvider, stethContract)

      const logger: Winston.Logger = Winston.createLogger({
        format: Winston.format.simple(),
        transports: [new Winston.transports.Console()],
      })

      const vaultWatcherSrv = new VaultWatcherSrv(logger, ethClient)
      const multisigWatcherSrv = new MultisigWatcherSrv(logger)

      App.instance = {
        ethClient: ethClient,
        VaultWatcherSrv: vaultWatcherSrv,
        MultisigWatcherSrv: multisigWatcherSrv,
        findingsRW: new DataRW([]),
        healthChecker: new HealthChecker(BorderTime, MaxNumberErrorsPerBorderTime),
      }
    }

    return App.instance
  }
}
