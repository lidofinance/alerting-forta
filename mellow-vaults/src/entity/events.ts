import { FindingSeverity, FindingType, Log } from 'forta-agent'

export type EventOfNotice = {
  name: string
  address: string
  event: string
  alertId: string
  description: CallableFunction
  severity: FindingSeverity
  type?: FindingType
  uniqueKey?: string
}

export type TransactionDtoTx = {
  data: string
  hash: string
}

export type TransactionDto = {
  logs: Log[]
  to: string | null
  block: {
    timestamp: number
    number: number
  }
  transaction: TransactionDtoTx
}

export type BlockDto = {
  number: number
  timestamp: number
  parentHash: string
}
