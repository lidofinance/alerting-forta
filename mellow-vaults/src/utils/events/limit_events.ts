import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'

export const MELLOW_VAULT_INCREASE_LIMIT_EVENT = 'event IncreaseLimit(uint256 amount)'

export const vaultLimitNotice = {
  IncreaseLimit: {
    event: MELLOW_VAULT_INCREASE_LIMIT_EVENT,
    alertId: 'MELLOW-SYMBIOTIC-INCREASE-LIMIT',
    name: '⚠️ Vault: Symbiotic limit increased',
    description: (args: Result) => `Symbiotic limit increased to ${args.amount}`,
    severity: FindingSeverity.Medium,
    type: FindingType.Info,
    uniqueKey: '27c9ac49-4c5f-4d3e-9a4a-0d61e21965b2 ',
  },
}
