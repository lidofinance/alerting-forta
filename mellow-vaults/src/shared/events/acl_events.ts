import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { ACL_ROLES } from 'constants/common'

export function getACLEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = [
    '12c18264-900b-48a9-bdee-b193b61a36c6',
    'd45a13a5-940c-439f-96e2-2cec78f9de24',
    '19ba7243-ff94-45af-aa3d-6b12a7f0b83d',
  ]

  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event:
        'event RoleAdminChanged(bytes32 indexed role, bytes32 indexed previousAdminRole, bytes32 indexed newAdminRole)',
      alertId: 'MELLOW-VAULT-ROLE-ADMIN-CHANGED',
      name: 'ðŸš¨ Vault: Role Admin changed',
      description: (args: Result) =>
        `Role Admin for role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was changed from ${args.previousAdminRole} to ${args.newAdminRole} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event RoleGranted(bytes32 indexed role, address indexed account, address indexed sender)',
      alertId: 'MELLOW-VAULT-ROLE-GRANTED',
      name: 'ðŸš¨ Vault: Role granted',
      description: (args: Result) =>
        `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was granted to ${args.account} by ${args.sender} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[1],
    },
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender)',
      alertId: 'MELLOW-VAULT-ROLE-REVOKED',
      name: 'ðŸš¨  Vault: Role revoked',
      description: (args: Result) =>
        `Role ${args.role}(${ACL_ROLES.get(args.role) || 'unknown'}) ` +
        `was revoked to ${args.account} by ${args.sender} ` +
        `bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`,
      severity: FindingSeverity.High,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[2],
    },
  ]
}
