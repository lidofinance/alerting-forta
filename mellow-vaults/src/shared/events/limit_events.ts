import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { Result } from '@ethersproject/abi/lib'
import { MELLOW_SYMBIOTIC_ADDRESS } from 'constants/common'

export function getLimitEvents(): EventOfNotice[] {
  const uniqueKeys = ['27c9ac49-4c5f-4d3e-9a4a-0d61e21965b2 ']

  return [
    {
      address: MELLOW_SYMBIOTIC_ADDRESS,
      event: 'event IncreaseLimit(uint256 amount)',
      alertId: 'MELLOW-SYMBIOTIC-INCREASE-LIMIT',
      name: '⚠️ Vault: Symbiotic limit increased',
      description: (args: Result) => `Symbiotic limit increased to ${args.amount}`,
      severity: FindingSeverity.Medium,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
  ]
}
