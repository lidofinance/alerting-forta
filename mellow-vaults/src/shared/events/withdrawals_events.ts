import { EventOfNotice } from '../../entity/events'
import { FindingSeverity, FindingType } from 'forta-agent'
import { VAULT_LIST } from 'constants/common'

export function getProcessAllEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = ['9d846c64-6858-47c4-afd2-3f3d31da079b']

  const vault = VAULT_LIST.find((vault) => vault.defaultBondStrategy === DEFAULT_BOND_STRATEGY_ADDRESS)
  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event DefaultBondStrategyProcessWithdrawals (address[] users, uint256 timestamp)',
      alertId: 'MELLOW-VAULT-WITHDRAWAL-ALL',
      name: 'ℹ️ Vault: Withdrawal all',
      description: () => {
        return `Mellow Vault [${vault?.name}] (${vault?.vault}) - withdrawal, bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`
      },
      severity: FindingSeverity.Info,
      type: FindingType.Info,
      uniqueKey: uniqueKeys[0],
    },
  ]
}

export function getProcessWithdrawalsEvents(DEFAULT_BOND_STRATEGY_ADDRESS: string): EventOfNotice[] {
  const uniqueKeys = ['e27d8c35-bce0-4901-a3d6-ca232cf68ede']

  const vault = VAULT_LIST.find((vault) => vault.defaultBondStrategy === DEFAULT_BOND_STRATEGY_ADDRESS)
  return [
    {
      address: DEFAULT_BOND_STRATEGY_ADDRESS,
      event: 'event DefaultBondStrategyProcessWithdrawals (address[] users, uint256 timestamp)',
      alertId: 'MELLOW-VAULT-PARTIAL-WITHDRAWAL',
      name: '⚠️ Vault: Withdrawal partial',
      description: () => {
        return `Mellow Vault [${vault?.name}] (${vault?.vault}) - bond strategy address - ${DEFAULT_BOND_STRATEGY_ADDRESS}`
      },
      severity: FindingSeverity.Medium,
      type: FindingType.Suspicious,
      uniqueKey: uniqueKeys[0],
    },
  ]
}
