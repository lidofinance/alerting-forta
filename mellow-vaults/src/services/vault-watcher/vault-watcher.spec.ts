import { BlockDto, TransactionDto } from '../../entity/events'
import * as Winston from 'winston'
import { ethers, Finding } from 'forta-agent'

import { VaultWatcherSrv } from './VaultWatcher.srv'
import { App } from '../../app'

const TEST_TIMEOUT = 120_000 // ms

describe('VaultWatchers srv functional tests', () => {
  const drpcURL = `https://eth.drpc.org`
  const mainnet = 1
  const ethProvider = new ethers.providers.JsonRpcProvider(drpcURL, mainnet)

  const ethClient = App.prepareClient(ethProvider)

  const logger: Winston.Logger = Winston.createLogger({
    format: Winston.format.simple(),
    transports: [new Winston.transports.Console()],
  })

  const service = new VaultWatcherSrv(logger, ethClient)

  let runTransaction: (txHash: string, initBlock?: number) => Promise<Finding[]>
  let runBlock: (blockHashOrNumber: string | number, initBlock: number) => Promise<Finding[]>

  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation(() => {})
    jest.spyOn(Date, 'now').mockImplementation(() => new Date('2024-01-01').getTime())

    runBlock = async (blockHashOrNumber, initBlock) => {
      if (initBlock > Number(blockHashOrNumber)) {
        throw new Error('Wrong initial block or argument order')
      }
      const initErr = await service.initialize(initBlock)
      if (initErr instanceof Error) {
        throw initErr
      }
      const block = await ethProvider.getBlock(blockHashOrNumber)
      const blockDto: BlockDto = {
        number: block.number,
        timestamp: block.timestamp,
        parentHash: block.parentHash,
      }

      return service.handleBlock(blockDto)
    }

    runTransaction = async (txHash: string) => {
      const trx = await ethProvider.getTransaction(txHash)
      const receipt = await trx.wait()

      const transactionDto: TransactionDto = {
        logs: receipt.logs,
        to: trx.to ? trx.to : null,
        block: {
          timestamp: trx.timestamp ? trx.timestamp : new Date().getTime(),
          number: trx.blockNumber ? trx.blockNumber : 1,
        },
        transaction: {
          data: trx.data || '',
          hash: trx.hash || '',
        },
      }

      return service.handleTransaction(transactionDto)
    }
  })

  afterAll(() => {
    jest.resetAllMocks()
  })

  it(
    'should RoleRevoked',
    async () => {
      const findings = await runTransaction('0xb8b96ee47cadb80ec41dafceba1fe5b10b1d50b75e7ad9114ee16180bda2d2b4')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should RoleGranted',
    async () => {
      const findings = await runTransaction('0x28cbf0521b4e6c03bf719cf958516609b3bf678b2f20dc958ae2db8999345fc5')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should RoleAdminChanged (3+2)',
    async () => {
      const findings = await runTransaction('0x047e8dd86f2e58fd05630e3fa1d1b1ce4a7638e3138cf9762894cf99a4fcabe6')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should be not slot changes',
    async () => {
      const findings = await runTransaction(
        '0xffef3e8046d882ccd5773582438ebf6576e4093dec734ed27e0f5065d1a89d7b',
        20061165,
      )
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should everything is fine',
    async () => {
      const findings = await runBlock(20207135, 20207134)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should process tx with Added Owner and Execution Success',
    async () => {
      const findings = await runTransaction('0x4c3e507fd78daac24dd75e33e7c69381481db8bce62c795d2eec67126ad6d396')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should show limit inc',
    async () => {
      const findings = await runTransaction('0xfb1a547bd1da10635513779fc7ecfd12d8db3f39593f7538ab77f61484592a3c')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should process tx with Changed Threshold and Execution Success',
    async () => {
      const findings = await runTransaction('0xb58e9e81ad1dac1f33b9dfc4d19f2d909a2a3ea890c31aa27a0df10f86bd4eea')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should withdrawal from contract',
    async () => {
      const findings = await runTransaction('0x17c222325fa85abc5a0e74708264c751cbf54e6f06fdfd852d520e3bee1b0596')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should withdrawal from curator',
    async () => {
      const findings = await runTransaction('0x07f911ae589572a8f73404289932e9faf70735eaa9347c79557888dd10624fb5')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should withdrawal from curator multicall',
    async () => {
      const findings = await runTransaction('0x5fb0b597627bfc2c4031738f99ee28df7e831b8ca9277cc4d5fe62e16506ce51')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should should find 1 withdrawal',
    async () => {
      const findings = await runBlock(20203686, 20203685)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find no-withdrawals at 1 vaults',
    async () => {
      const findings = await runBlock(20188825, 20188824)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should process tx with Re7 Vault withdrawal all',
    async () => {
      const findings = await runTransaction('0x74ba54139103dc266b50cb6e2f04d53398809290851a4c90d8d40e73a4db5e8e')
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find Steakhouse Vault slot change',
    async () => {
      const findings = await runBlock(20266488, 20266480)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )

  it(
    'should find InfStones Restaked slot change',
    async () => {
      const findings = await runBlock(20184204, 20184203)
      expect(findings).toMatchSnapshot()
    },
    TEST_TIMEOUT,
  )
})
