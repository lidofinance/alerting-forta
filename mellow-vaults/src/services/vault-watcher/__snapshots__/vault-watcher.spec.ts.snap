// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaultWatchers srv functional tests should RoleAdminChanged (3+2) 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-GRANTED",
    "description": "Mellow Vault [Steakhouse Vault] bond strategy Role 0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622(OPERATOR) was granted to 0x188858AC61a74350116d1CB6958fBc509FD6afA1 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622,0x188858AC61a74350116d1CB6958fBc509FD6afA1,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö® Vault: Role granted",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "d45a13a5-940c-439f-96e2-2cec78f9de24",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-GRANTED",
    "description": "Mellow Vault [Steakhouse Vault] bond strategy Role 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8(ADMIN ROLE) was granted to 0x188858AC61a74350116d1CB6958fBc509FD6afA1 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8,0x188858AC61a74350116d1CB6958fBc509FD6afA1,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö® Vault: Role granted",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "d45a13a5-940c-439f-96e2-2cec78f9de24",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-ADMIN-CHANGED",
    "description": "Mellow Vault [Steakhouse Vault] bond strategy Role Admin for role 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8(ADMIN ROLE) was changed from 0x0000000000000000000000000000000000000000000000000000000000000000 to 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8,0x0000000000000000000000000000000000000000000000000000000000000000,0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8",
    },
    "name": "üö® Vault: Role Admin changed",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "12c18264-900b-48a9-bdee-b193b61a36c6",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-ADMIN-CHANGED",
    "description": "Mellow Vault [Steakhouse Vault] bond strategy Role Admin for role 0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7(ADMIN DELEGATE ROLE) was changed from 0x0000000000000000000000000000000000000000000000000000000000000000 to 0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7,0x0000000000000000000000000000000000000000000000000000000000000000,0xf23ec0bb4210edd5cba85afd05127efcd2fc6a781bfed49188da1081670b22d8",
    },
    "name": "üö® Vault: Role Admin changed",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "12c18264-900b-48a9-bdee-b193b61a36c6",
  },
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-ADMIN-CHANGED",
    "description": "Mellow Vault [Steakhouse Vault] bond strategy Role Admin for role 0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622(OPERATOR) was changed from 0x0000000000000000000000000000000000000000000000000000000000000000 to 0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622,0x0000000000000000000000000000000000000000000000000000000000000000,0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7",
    },
    "name": "üö® Vault: Role Admin changed",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "12c18264-900b-48a9-bdee-b193b61a36c6",
  },
]
`;

exports[`VaultWatchers srv functional tests should RoleGranted 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-GRANTED",
    "description": "Mellow Vault [Steakhouse Vault] bond strategy Role 0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622(OPERATOR) was granted to 0x2afc096981c2CFe3501bE4054160048718F6C0C8 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0x46a52cf33029de9f84853745a87af28464c80bf0346df1b32e205fc73319f622,0x2afc096981c2CFe3501bE4054160048718F6C0C8,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö® Vault: Role granted",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "d45a13a5-940c-439f-96e2-2cec78f9de24",
  },
]
`;

exports[`VaultWatchers srv functional tests should RoleRevoked 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-ROLE-REVOKED",
    "description": "Mellow Vault [Steakhouse Vault] bond strategy Role 0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7(ADMIN DELEGATE ROLE) was revoked to 0x188858AC61a74350116d1CB6958fBc509FD6afA1 by 0x188858AC61a74350116d1CB6958fBc509FD6afA1 bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xc171260023d22a25a00a2789664c9334017843b831138c8ef03cc8897e5873d7,0x188858AC61a74350116d1CB6958fBc509FD6afA1,0x188858AC61a74350116d1CB6958fBc509FD6afA1",
    },
    "name": "üö®  Vault: Role revoked",
    "protocol": "ethereum",
    "severity": 4,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "19ba7243-ff94-45af-aa3d-6b12a7f0b83d",
  },
]
`;

exports[`VaultWatchers srv functional tests should be not slot changes 1`] = `[]`;

exports[`VaultWatchers srv functional tests should everything is fine 1`] = `[]`;

exports[`VaultWatchers srv functional tests should find InfStones Restaked slot change 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-STORAGE-SLOT-VALUE-CHANGED",
    "description": "Mellow Vault [InfStones Restaked ETH Vault] Value of the storage slot \`'maximalTotalSupply'\` for contract 0x49cd586dd9ba227be9654c735a659a1db08232a9 has changed!
Prev value: 1000000000000000000
New value: 20000000000000000000000",
    "labels": [],
    "metadata": {},
    "name": "üö®üö®üö® Vault critical storage slot value changed",
    "protocol": "ethereum",
    "severity": 5,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
]
`;

exports[`VaultWatchers srv functional tests should find Steakhouse Vault slot change 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-STORAGE-SLOT-VALUE-CHANGED",
    "description": "Mellow Vault [Steakhouse Vault] Value of the storage slot \`'maximalTotalSupply'\` for contract 0xbeef69ac7870777598a04b2bd4771c71212e6abc has changed!
Prev value: 16000000000000000000000
New value: 36000000000000000000000",
    "labels": [],
    "metadata": {},
    "name": "üö®üö®üö® Vault critical storage slot value changed",
    "protocol": "ethereum",
    "severity": 5,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
]
`;

exports[`VaultWatchers srv functional tests should find no-withdrawals at Re7 vault 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "VAULT-NO-WITHDRAWAL-48",
    "description": "Mellow Vault [Re7 Vault] (0x84631c0d0081fde56deb72f6de77abbbf6a9f93a) pending withdrawers count - 6",
    "labels": [],
    "metadata": {},
    "name": "‚ö†Ô∏è Vault: Withdrawals haven‚Äôt been called for at least 48 hours",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 2,
    "uniqueKey": "",
  },
]
`;

exports[`VaultWatchers srv functional tests should not find no-withdrawals at Steakhouse vault 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "VAULT-WSTETH-LIMIT-REACHED",
    "description": "Symbiotic - 0xc329400492c6ff2438472d4651ad17389fcb843a",
    "labels": [],
    "metadata": {},
    "name": "‚ö†Ô∏è Vault: Symbiotic limit reached",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "",
  },
]
`;

exports[`VaultWatchers srv functional tests should process tx with Added Owner and Execution Success 1`] = `[]`;

exports[`VaultWatchers srv functional tests should process tx with Changed Threshold and Execution Success 1`] = `[]`;

exports[`VaultWatchers srv functional tests should process tx with Re7 Vault withdrawal all 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-WITHDRAWAL-ALL",
    "description": "Mellow Vault [Re7 Vault] (0x84631c0d0081fde56deb72f6de77abbbf6a9f93a) - withdrawal, bond strategy address - 0xce3a8820265ad186e8c1ceaed16ae97176d020ba",
    "labels": [],
    "metadata": {
      "args": "0x15C273AA5aDd5433D5e806a3DB968A8656797fbA,0x61ae94dd817357B59413202Ac60B5997192AAE0F,0x520F977ff2fb5fc3a188B9B7821d583BC8741BC7,0x7Bd350bb79fD41d9F95E39005FcD973DD9d9a087,0xc4899b8AA371917e382E989078Ce22469B822f18,0x9fb1750Da6266a05601855bb62767eBC742707B1,1719575915",
    },
    "name": "‚ÑπÔ∏è Vault: Withdrawal all",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "9d846c64-6858-47c4-afd2-3f3d31da079b",
  },
]
`;

exports[`VaultWatchers srv functional tests should should find 1 withdrawal 1`] = `[]`;

exports[`VaultWatchers srv functional tests should show limit inc 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-SYMBIOTIC-INCREASE-LIMIT",
    "description": "Symbiotic limit increased to 169313200000000000000000",
    "labels": [],
    "metadata": {
      "args": "169313200000000000000000",
    },
    "name": "‚ö†Ô∏è Vault: Symbiotic limit increased",
    "protocol": "ethereum",
    "severity": 3,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "27c9ac49-4c5f-4d3e-9a4a-0d61e21965b2 ",
  },
]
`;

exports[`VaultWatchers srv functional tests should withdrawal from contract 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-WITHDRAWAL-ALL",
    "description": "Mellow Vault [Steakhouse Vault] (0xbeef69ac7870777598a04b2bd4771c71212e6abc) - withdrawal, bond strategy address - 0x7a14b34a9a8ea235c66528dc3bf3aefc36dfc268",
    "labels": [],
    "metadata": {
      "args": "0xE9229230D3768b42e7c2a54cffbfE0f0288261aC,0x7fF11ad66Ec323C6A5D40363946De553D8cf2B2E,0x1A63572CF23505DDBCCB2Bdd0cE516Be9f98E842,0x21ded9833e786Ea941410312e7e78Fc79fc879Af,0x9A3b9ae785C399bcdb04115fe78ad866EE39c877,1718609063",
    },
    "name": "‚ÑπÔ∏è Vault: Withdrawal all",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "9d846c64-6858-47c4-afd2-3f3d31da079b",
  },
]
`;

exports[`VaultWatchers srv functional tests should withdrawal from curator 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-WITHDRAWAL-ALL",
    "description": "Mellow Vault [Mev Capital Vault] (0x5fd13359ba15a84b76f7f87568309040176167cd) - withdrawal, bond strategy address - 0xc3a149b5ca3f4a5f17f5d865c14aa9dbb570f10a",
    "labels": [],
    "metadata": {
      "args": "0xf41BF1334ea7399ae21F94a434AE8B5486f9dF37,0x737E37dfcCAd5B6bEf491202072170290EcAA1aF,1718890811",
    },
    "name": "‚ÑπÔ∏è Vault: Withdrawal all",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "9d846c64-6858-47c4-afd2-3f3d31da079b",
  },
]
`;

exports[`VaultWatchers srv functional tests should withdrawal from curator multicall 1`] = `
[
  Finding {
    "addresses": [],
    "alertId": "MELLOW-VAULT-WITHDRAWAL-ALL",
    "description": "Mellow Vault [Re7 Vault] (0x84631c0d0081fde56deb72f6de77abbbf6a9f93a) - withdrawal, bond strategy address - 0xce3a8820265ad186e8c1ceaed16ae97176d020ba",
    "labels": [],
    "metadata": {
      "args": "0x718E7B7e03F9370394CC8a3BD41B395C72B90fa2,0x3DA0C44465d7b662208Fe6216d17B7f535A472Ad,0x09B0dfBBF2e6d98FB665b9694DBcA7b08193db4e,1718960795",
    },
    "name": "‚ÑπÔ∏è Vault: Withdrawal all",
    "protocol": "ethereum",
    "severity": 1,
    "source": {},
    "timestamp": 2024-01-01T00:00:00.000Z,
    "type": 4,
    "uniqueKey": "9d846c64-6858-47c4-afd2-3f3d31da079b",
  },
]
`;
